libnmstate/nm/checkpoint.py

```diff
@@ -92,21 +47,122 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+                self.cp.get_path(),
+                self.mainloop.cancellable,
+                _checkpoint_destroy_callback,
+                userdata
+            )
+            if not self.mainloop.run(self._timeout):
+                raise CheckPointError("Failed to destroy checkpoint: "
+                                      "{}".format(self.mainloop.error))
+        else:
+            raise CheckPointError('No checkpoint to destroy')
 
     def rollback(self):
-        result = self._manager.interface.CheckpointRollback(self._checkpoint)
-        logging.debug('Checkpoint %s rollback executed: %s',
-                      self._checkpoint, result)
-        return result
+        if self.cp:
+            userdata = self
+            self.mainloop.push_action(
+                self._nmclient.checkpoint_rollback,
+                self.cp.get_path(),
+                self.mainloop.cancellable,
+                _checkpoint_rollback_callback,
+                userdata
+            )
+            if not self.mainloop.run(self._timeout):
+                raise CheckPointError("Failed to rollback checkpoint: "
+                                      "{}".format(self.mainloop.error))
+        else:
+            raise CheckPointError('No checkpoint to rollback')
+
+
+def _checkpoint_create_callback(client, result, data):
+    checkpoint = data
+    mainloop = checkpoint.mainloop
+    try:
+        cp = client.checkpoint_create_finish(result)
+        checkpoint.cp = cp
+        if cp:
+            logging.debug('Checkpoint {} created for all devices'.format(
+                           cp.get_path()))
+            mainloop.execute_next_action()
+        else:
+            mainloop.quit('Checkpoint creation failed: error={}'.format(
+                          mainloop.error)
+    except Exception as e:
+        if mainloop.is_action_canceled(e):
+            logging.debug('Checkpoint creation canceled: error={}'.format(e))
+        else:
+            mainloop.quit('Checkpoint creation failed: error={}'.format(e))
+
+
+def _checkpoint_rollback_callback(client, result, data):
+    checkpoint = data
+    mainloop = checkpoint.mainloop
+    cp = checkpoint.cp
+    try:
+        _check_rollback_result(client, result, cp)
+        checkpoint.cp = None
+        mainloop.execute_next_action()
+    except Exception as e:
+        if mainloop.is_action_canceled(e):
+            logging.debug('Checkpoint rollback canceled: error={}'.format(e))
+        else:
+            mainloop.quit('Checkpoint rollback failed: error={}'.format(e))
+
+
+def _check_rollback_result(client, result, cp):
+    ret = client.checkpoint_rollback_finish(result)
+    logging.debug('Checkpoint %s rollback executed', cp.get_path())
+    for path in ret:
+        d = client.get_device_by_path(path)
+        if d is None:
+            iface = path
+        else:
+            iface = d.get_iface()
+        logging.debug('Interface %s rollback succeeded', iface)
```

Comment:
This should be under the `else` of the following condition, right?
Perhaps use `if ret[path] == 0` if this is indeed the case.

Reply:
Done.